typing/typecore.ml

```diff
@@ -1803,108 +1803,107 @@ let rec is_nonexpansive exp =
-      Vars.fold (fun _ (mut,_,_) b -> decr count; b && mut = Immutable)
-        vars true &&
-      !count = 0
+          | Tcf_initializer e -> is_expansive e
+          | Tcf_constraint _ -> false
+          | Tcf_inherit _ -> true
+          | Tcf_attribute _ -> false)
+        fields ||
+      Vars.fold (fun _ (mut,_,_) b -> decr count; b || mut = Mutable)
+        vars false ||
+      !count <> 0
   | Texp_letmodule (_, _, _, mexp, e)
   | Texp_open ({ open_expr = mexp; _}, e) ->
-      is_nonexpansive_mod mexp && is_nonexpansive e
+      is_expansive_mod mexp || is_expansive e
   | Texp_pack mexp ->
-      is_nonexpansive_mod mexp
+      is_expansive_mod mexp
   (* Computations which raise exceptions are nonexpansive, since (raise e) is
      equivalent to (raise e; diverge), and a nonexpansive "diverge" can be
      produced using lazy values or the relaxed value restriction.
      See GPR#1142 *)
   | Texp_assert exp ->
-      is_nonexpansive exp
+      is_expansive exp
   | Texp_apply (
       { exp_desc = Texp_ident (_, _, {val_kind =
              Val_prim {Primitive.prim_name =
                          ("%raise" | "%reraise" | "%raise_notrace")}}) },
       [Nolabel, Some e]) ->
-     is_nonexpansive e
-  | _ -> false
+     is_expansive e
+  | _ -> true
 
-and is_nonexpansive_mod mexp =
+and is_expansive_mod mexp =
   match mexp.mod_desc with
-  | Tmod_ident _ -> true
-  | Tmod_functor _ -> true
-  | Tmod_unpack (e, _) -> is_nonexpansive e
-  | Tmod_constraint (m, _, _, _) -> is_nonexpansive_mod m
+  | Tmod_ident _
+  | Tmod_functor _ -> false
+  | Tmod_unpack (e, _) -> is_expansive e
+  | Tmod_constraint (m, _, _, _) -> is_expansive_mod m
   | Tmod_structure str ->
-      List.for_all
+      List.exists
         (fun item -> match item.str_desc with
           | Tstr_eval _ | Tstr_primitive _ | Tstr_type _
-          | Tstr_modtype _ | Tstr_class_type _  -> true
+          | Tstr_modtype _ | Tstr_class_type _  -> false
           | Tstr_value (_, pat_exp_list) ->
-              List.for_all (fun vb -> is_nonexpansive vb.vb_expr) pat_exp_list
+              List.exists (fun vb -> is_expansive vb.vb_expr) pat_exp_list
           | Tstr_module {mb_expr=m;_}
           | Tstr_open {open_expr=m;_}
-          | Tstr_include {incl_mod=m;_} -> is_nonexpansive_mod m
+          | Tstr_include {incl_mod=m;_} -> is_expansive_mod m
           | Tstr_recmodule id_mod_list ->
-              List.for_all (fun {mb_expr=m;_} -> is_nonexpansive_mod m)
+              List.exists (fun {mb_expr=m;_} -> is_expansive_mod m)
                 id_mod_list
-          | Tstr_exception {tyexn_constructor = {ext_kind = Text_decl _}} ->
-              false (* true would be unsound *)
-          | Tstr_exception {tyexn_constructor = {ext_kind = Text_rebind _}} ->
-              true
+          | Tstr_exception
+              {tyexn_constructor = {ext_kind = (Text_decl _ | Text_rebind _) }}
+            ->
+              true (* false would be unsound *)
```

Comment:
Seems like the `Text_rebind` case is also not right.

Reply:
Indeed. Fixed.